aten/src/ATen/native/cuda/Sort.cpp

```diff
@@ -0,0 +1,182 @@
+#include <ATen/native/cuda/Sort.h>
+#include <ATen/ExpandUtils.h>
+#include <ATen/Functions.h>
+#include <ATen/NativeFunctions.h>
+#include <ATen/MemoryOverlap.h>
+#include <ATen/WrapDimUtils.h>
+#include <ATen/native/Resize.h>
+
+#include <limits>
+
+namespace at { namespace native {
+
+// We perform a segmented sort in cub with inputs that have
+// more than 1024/2048 elements along the selected dimension.
+// Otherwise, we do an inplace bitonic sort (see sortKeyValueInplace).
+bool should_use_small_sort(const TensorBase &self, int64_t dim) {
+  int64_t ndim = self.dim();
+  int64_t nsort = self.sizes()[dim];
+  int64_t threshold;
+  if (self.scalar_type() == kLong || self.scalar_type() == kDouble) {
+    threshold = 1024;
+  } else {
+    threshold = 2048;
+  }
+  return nsort <= threshold;
+}
+
+std::vector<int64_t> infer_dense_strides_dim_last(const Tensor & self, int64_t dim);
```

Comment:
I think it might be related to this comment:
https://github.com/pytorch/pytorch/blob/b00206d4737d1f1e7a442c9f8a1cadccd272a386/aten/src/ATen/native/cuda/UniqueCub.cu#L119-L121

`SortImpl.cu` includes thrust and `Sort.cu` includes cub. Although since this is a host-only function, it would be better to not use `thrust` at all. I'll leave that for a separate PR though.

Reply:
thanks!